cachi2/core/package_managers/pip.py

```diff
@@ -1865,14 +1895,21 @@ def _download_from_requirement_files(
         reference
     :raises PackageRejected: If requirement file does not exist
     """
-    requirements = []
-    for req_file in files:
-        if not req_file.path.exists():
+    req_files = []
+    for file in files:
+        if not file.path.exists():
             raise PackageRejected(
-                f"The requirements file does not exist: {req_file}",
+                f"The requirements file does not exist: {file}",
                 solution="Please check that you have specified correct requirements file paths",
             )
-        requirements.extend(_download_dependencies(output_dir, PipRequirementsFile(req_file)))
+        req_files.append(PipRequirementsFile(file))
+
+    checksums = _get_user_checksums(req_files, allow_binary)
```

Comment:
Thread on "handle checksums on repository level vs. single file level"

#### Scenario 1

let's say that
* project_a/requirements.txt has `requests==2.28.2 --hash=<h1> --hash=<h2>`
* project_b/requirements.txt has `requests==2.28.2` with no hash options

requests==2.28.2 has a total of 2 archives (1 wheel, 1 sdist), so this just means that the downloaded dependencies may be unverified even though project_a did intend to verify them.

Whether the hashes get verified or not depends on
* do we re-download things we already downloaded? (currently that's yes for sdists, no for wheels)
* does project_a or project_b get processed first?

This part could more or less be solved with the `cachi2:missing_hash:in_file` properties to report the potential issue. For what it's worth, I believe every package manager we support currently has this issue.

Reply:
#### Scenario 2

* project_c/requirements.txt has `charset-normalizer==3.2.0 --hash=<h1> --hash=<h2>`
* project_d/requirements.txt has `charset-normalizer==3.2.0 --hash=<h1> --hash=<h3>`

charset-normalizer==3.2.0 has a sdist and 74 wheels. Each project has the hashes for the sdist and exactly 1 wheel. Maybe they use different python versions and therefore need a different wheel?

In this case, we need to ensure we download both the wheel which matches h2 and the one which matches h3.

I believe the PR in its current state handles this properly (though very inefficiently in case that the PyPI server does not report the expected hashes)

Reply:
#### Scenario 3

* project_c/requirements.txt has `charset-normalizer==3.2.0 --hash=<h1> --hash=<h2>`
* project_d/requirements.txt has `charset-normalizer==3.2.0` with no hashes

In this case, we download the sdist and all 74 wheels, but 73 of the wheels don't get checksum-verified.

What's worse, one of the 73 wheels can accidentally be picked up by project_c (if pip determines that this wheel is a better match for the user's platform than the one which matches h2). This is partially true for Scenario 2 as well, but less likely.

But again, maybe `cachi2:missing_hash:in_file` is good enough here.

Reply:
So IMO we can more or less disregard Scenario 2, because it's handle-able on single requirements-file level as well.

When we were discussing it on the "Team PR review", I didn't realize that we handle it properly because we don't re-download, re-verify and delete-on-mismatch wheel files that were already downloaded.

Reply:
And I think it really comes down to "is reporting the missing hashes good enough?". For pip it's worse than for other package managers, but maybe good enough.

Reply:
@brunoapimentel @taylormadore @lkolacek @ejegrova @ben-alkov @fepas 

What do you think?

Handling it repo-wide would be more secure, but probably surprising to users. And it kinda just feels weird.

I was also looking at how `pip` itself handles it for inspiration, but... let's not get inspired by pip :sweat_smile: 
- if file_a has hashes and file_b doesn't, pip will verify hashes
- if file_a does't and file_b does, pip will fail
- if file_a and file_b have different hashes, pip will only look at file_a

Reply:
So, in scenario 1, the wheel could be checksum verified or not dependending on the order of the packages specified in the CLI, right? I'm wondering if any user would find it odd that that specific file is marked as `missing_hash`, even though it could have its checksum verified.

I do agree that the repo-wide approach feels weird, though, so I'm still leaning towards looking it at a single file level.

Reply:
Yeah, and what's even more unfortunate - the file that doesn't have hashes can prevent checksum verification for the one that does (if you download all the wheels from that file first, you don't try to re-download and re-verify them later). But still, if EC blocks something based on missing hashes, that might be OK

Reply:
Maybe we could workaround having the order of input packages impact the checksum result. What if we remove the check for the file existence from [here](https://github.com/slimreaper35/cachi2/blob/STONEBLD-1332/cachi2/core/package_managers/pip.py#L1404), and only consider it for the files that'll be downloaded [here](https://github.com/slimreaper35/cachi2/blob/STONEBLD-1332/cachi2/core/package_managers/pip.py#L1434)? That is, leave existing files in the `to_download` list, but skip downloading them when passing the actual list to the asyncio_download function.

This way, we'd at least always verify the checksums of files that relate to an input package. Then, I think it's a matter for the PropertySet deduplication to consider that a file had its hashes verified if at least one of the reported SBOM components says it was.

Wdyt?

Reply:
What would you do if an already-downloaded wheel does not match any checksum in the next requirements file? Keep in mind that there's also [Scenario 2](https://github.com/containerbuildsystem/cachi2/pull/255#discussion_r1297342998).

Reply:
You're right, that would make scenario 2 fail. I guess we can file a story so we can look at the possibility of handling this repo-wide, but let's leave this out of this PR. Although it's a corner case, the thought of the output changing depending on the order of packages disturbs me :smiling_face_with_tear:

Reply:
For what it's worth, this PR is already halfway there to handling it repo-wide. We would just have to collect the `checksums` from _all_ requirements files, not just the requirements files of a single project as we do here https://github.com/containerbuildsystem/cachi2/pull/255/files#diff-cfda53cfc3e45684cd3ac813162ad9ee144594345ff841a69250cc88de960dbcR1901

But we would most likely want to change the behavior of [Scenario 3](https://github.com/containerbuildsystem/cachi2/pull/255#discussion_r1297343172) then - download only the two archives that match the hashes, not all 74 of them. Or download the matching wheels first to make sure those two are checksum-verified and only then download the remaining 72

Reply:
But to summarize: let's handle it on single-file level now and file an story/investigation for how to handle it better. 

Sound good?

Reply:
so @slimreaper35 this basically means we don't need to combine the checksums from different requirements files like we do with `_get_user_checksums` (at least for now)

Reply:
If it's easier to remove it than to keep it, I'd say yes. Otherwise we might as well keep it, it could be useful for the future improvement.